File: leetcode/easy/two-sum-ii-input-array-is-sorted/dduck-two-sum-ii-input-array-is-sorted.js
// https://leetcode.com/problems/two-sum-ii-input-array-is-sorted/
// Runtime: 48 ms, faster than 98.15% of JavaScript online submissions
// Memory Usage: 38.4 MB, less than 8.70% of JavaScript online submissions
// Runtime: 56 ms, faster than 78.76% of JavaScript online submissions
// Memory Usage: 38.5 MB, less than 8.70% of JavaScript online submissions

const twoSum = (numbers, target, leftIndex = 0, rightIndex = numbers.length - 1) => {
  const sum = numbers[leftIndex] + numbers[rightIndex];
  if (sum === target) {
    return [leftIndex + 1, rightIndex + 1];
  }
  if (sum > target) {
    rightIndex -= 1;
  } else {
    leftIndex += 1;
  }
  return twoSum(numbers, target, leftIndex, rightIndex);
};

const numbers = [2, 7, 11, 15];
const target = 9;
console.log(twoSum(numbers, target), [1, 2]);
// Input: numbers = [2,7,11,15], target = 9
// Output: [1,2]
// Explanation: The sum of 2 and 7 is 9. Therefore index1 = 1, index2 = 2.

File: leetcode/easy/167_two_sum_ii_input_array_is_sorted/dduck-two-sum-ii-input-array-is-sorted.py
# https://leetcode.com/problems/two-sum-ii-input-array-is-sorted/

from unittest import TestCase
from typing import List


class Solution:
# Runtime: 76 ms, faster than 24.56% of Python3 online submissions for Two Sum II - Input array is sorted.
# Memory Usage: 42.5 MB, less than 5.80% of Python3 online submissions for Two Sum II - Input array is sorted.
    def twoSum(self, numbers: List[int], target: int, leftIndex=0, rightIndex=-1) -> List[int]:
        sum = numbers[leftIndex] + numbers[rightIndex]
        if sum == target:
            return [leftIndex + 1, len(numbers) + rightIndex + 1]
        if sum > target:
            rightIndex -= 1
        else:
            leftIndex += 1
        return self.twoSum(numbers, target, leftIndex, rightIndex)

# Runtime: 60 ms, faster than 88.83% of Python3 online submissions for Two Sum II - Input array is sorted.
# Memory Usage: 13.2 MB, less than 100.00% of Python3 online submissions for Two Sum II - Input array is sorted.
    def twoSum2(self, numbers: List[int], target: int, leftIndex=0, rightIndex=-1) -> List[int]:
        sum = numbers[leftIndex] + numbers[rightIndex]
        while sum != target:
            if sum > target:
                rightIndex -= 1
            else:
                leftIndex += 1
            sum = numbers[leftIndex] + numbers[rightIndex]
        return [leftIndex + 1, len(numbers) + rightIndex + 1]


TestCase().assertEqual(Solution().twoSum2([2, 7, 11, 15], 9), [1, 2])
